File: src/sorting/cpp/quick_sort/test/test_quick_sort.cpp
/**
 * @file test_quick_sort.cpp
 * @brief Unit tests for the quick sort algorithm.
 */

#include "quick_sort.h"
#include "gtest/gtest.h"

/// @brief Test sorting an empty array.
TEST(QuickSortTest, EmptyArray) {
  std::vector<int> arr;
  std::vector<int> expected;
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting a single element array.
TEST(QuickSortTest, SingleElement) {
  std::vector<int> arr{1};
  std::vector<int> expected{1};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting two elements in sorted order.
TEST(QuickSortTest, TwoElementsSorted) {
  std::vector<int> arr{1, 2};
  std::vector<int> expected{1, 2};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting two elements in reverse order.
TEST(QuickSortTest, TwoElementsReversed) {
  std::vector<int> arr{2, 1};
  std::vector<int> expected{1, 2};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting a reverse-sorted array.
TEST(QuickSortTest, DecreasingOrder) {
  std::vector<int> arr{5, 4, 3, 2, 1};
  std::vector<int> expected{1, 2, 3, 4, 5};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting an already sorted array.
TEST(QuickSortTest, IncreasingOrder) {
  std::vector<int> arr{1, 2, 3, 4, 5};
  std::vector<int> expected{1, 2, 3, 4, 5};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting an array in random order.
TEST(QuickSortTest, RandomOrder) {
  std::vector<int> arr{2, 5, 1, 4, 3};
  std::vector<int> expected{1, 2, 3, 4, 5};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting an array with duplicate values.
TEST(QuickSortTest, WithDuplicates) {
  std::vector<int> arr{3, 1, 4, 1, 5, 9, 2, 6, 5};
  std::vector<int> expected{1, 1, 2, 3, 4, 5, 5, 6, 9};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting an array with negative numbers.
TEST(QuickSortTest, WithNegativeNumbers) {
  std::vector<int> arr{-3, 1, -4, 1, 5};
  std::vector<int> expected{-4, -3, 1, 1, 5};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

/// @brief Test sorting an array where all elements are the same.
TEST(QuickSortTest, AllSameElements) {
  std::vector<int> arr{5, 5, 5, 5};
  std::vector<int> expected{5, 5, 5, 5};
  quick_sort(arr);
  EXPECT_EQ(expected, arr);
}

File: src/sorting/python/quick_sort/tests/test_quick_sort.py
"""Tests for quick sort algorithms."""

import unittest
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from quick_sort import quick_sort, quick_sort_three_way


class TestQuickSort(unittest.TestCase):
    """Test cases for the quick_sort function (2-way partitioning)."""

    def test_empty_array(self):
        """Sorting an empty array returns an empty array."""
        self.assertEqual(quick_sort([]), [])

    def test_single_element(self):
        """Sorting a single element array returns the same array."""
        self.assertEqual(quick_sort([1]), [1])

    def test_two_elements_sorted(self):
        """Sorting two already sorted elements."""
        self.assertEqual(quick_sort([1, 2]), [1, 2])

    def test_two_elements_reversed(self):
        """Sorting two elements in reverse order."""
        self.assertEqual(quick_sort([2, 1]), [1, 2])

    def test_increasing_order(self):
        """Sorting an already sorted array."""
        self.assertEqual(quick_sort([1, 2, 3, 4, 5]), [1, 2, 3, 4, 5])

    def test_decreasing_order(self):
        """Sorting a reverse-sorted array."""
        self.assertEqual(quick_sort([5, 4, 3, 2, 1]), [1, 2, 3, 4, 5])

    def test_random_order(self):
        """Sorting an array in random order."""
        self.assertEqual(quick_sort([2, 5, 1, 4, 3]), [1, 2, 3, 4, 5])

    def test_with_duplicates(self):
        """Sorting an array with duplicate values."""
        self.assertEqual(quick_sort([3, 1, 4, 1, 5, 9, 2, 6, 5]), [1, 1, 2, 3, 4, 5, 5, 6, 9])

    def test_with_negative_numbers(self):
        """Sorting an array containing negative numbers."""
        self.assertEqual(quick_sort([-3, 1, -4, 1, 5]), [-4, -3, 1, 1, 5])

    def test_all_same_elements(self):
        """Sorting an array where all elements are the same."""
        self.assertEqual(quick_sort([5, 5, 5, 5]), [5, 5, 5, 5])


class TestQuickSortThreeWay(unittest.TestCase):
    """Test cases for the quick_sort_three_way function (3-way partitioning)."""

    def test_empty_array(self):
        """Sorting an empty array returns an empty array."""
        self.assertEqual(quick_sort_three_way([]), [])

    def test_single_element(self):
        """Sorting a single element array returns the same array."""
        self.assertEqual(quick_sort_three_way([1]), [1])

    def test_two_elements_sorted(self):
        """Sorting two already sorted elements."""
        self.assertEqual(quick_sort_three_way([1, 2]), [1, 2])

    def test_two_elements_reversed(self):
        """Sorting two elements in reverse order."""
        self.assertEqual(quick_sort_three_way([2, 1]), [1, 2])

    def test_increasing_order(self):
        """Sorting an already sorted array."""
        self.assertEqual(quick_sort_three_way([1, 2, 3, 4, 5]), [1, 2, 3, 4, 5])

    def test_decreasing_order(self):
        """Sorting a reverse-sorted array."""
        self.assertEqual(quick_sort_three_way([5, 4, 3, 2, 1]), [1, 2, 3, 4, 5])

    def test_random_order(self):
        """Sorting an array in random order."""
        self.assertEqual(quick_sort_three_way([2, 5, 1, 4, 3]), [1, 2, 3, 4, 5])

    def test_with_duplicates(self):
        """Sorting an array with duplicate values."""
        self.assertEqual(quick_sort_three_way([3, 1, 4, 1, 5, 9, 2, 6, 5]), [1, 1, 2, 3, 4, 5, 5, 6, 9])

    def test_with_negative_numbers(self):
        """Sorting an array containing negative numbers."""
        self.assertEqual(quick_sort_three_way([-3, 1, -4, 1, 5]), [-4, -3, 1, 1, 5])

    def test_all_same_elements(self):
        """Sorting an array where all elements are the same."""
        self.assertEqual(quick_sort_three_way([5, 5, 5, 5]), [5, 5, 5, 5])


if __name__ == "__main__":
    unittest.main()
